src/Nri/Ui/Mark/V6.elm

```diff
@@ -69,72 +74,100 @@ viewWithOverlaps :
-                                            ]
-                                        ]
-                                        (String.Extra.toSentenceOxford names)
-                                    ]
-                                , segment []
-                                ]
-                       )
-                )
+    overlappingStyles segments
+        |> List.concatMap
+            (\( content, maybeLabel, styles ) ->
+                case maybeLabel of
+                    Nothing ->
+                        [ viewSegment content styles ]
+
+                    Just label ->
+                        [ label, viewSegment content styles ]
             )
-            ( Set.empty maybeStringSorter, [] )
-        |> Tuple.second
 
 
-ignoreRepeats : Set (Maybe String) -> List Mark -> List Mark
-ignoreRepeats lastMarks list =
-    List.filter (\x -> not (Set.memberOf lastMarks x.name)) list
+{-| Compute the styles required to mark the segments.
+
+You can use this if you require more control over the structure of how marked elements are laid out than `viewWithOverlaps` provides.
+
+The `Maybe (Html msg)` result is a label element that should be rendered before the current segment.
+
+-}
+overlappingStyles : List ( content, List Mark ) -> List ( content, Maybe (Html msg), List Style )
+overlappingStyles segments =
+    let
+        -- We can't detect the end of a span of marks until after we are past it!
+        -- So we need to go back to the last set of styles and add the ended styles there
+        updateEndRevStyles patchRevStyles endedMarks =
+            case ( patchRevStyles, endedMarks ) of
+                ( [], _ ) ->
+                    patchRevStyles
+
+                ( _, [] ) ->
+                    patchRevStyles
+
+                ( ( prevContent, prevLabel, prevStyles ) :: otherPrevStyles, _ ) ->
+                    ( prevContent, prevLabel, prevStyles ++ (tagAfterContent endedMarks :: List.concatMap .endStyles endedMarks) ) :: otherPrevStyles
+
+        { priorMarks, revStyles } =
+            List.foldl
+                (\( content, marks ) state ->
+                    let
+                        currentMarks =
+                            Set.fromList markSorter marks
+
+                        startedMarks =
+                            Set.dropIf (Set.memberOf state.priorMarks) currentMarks
+                                |> Set.toList
+
+                        endedMarks =
+                            Set.dropIf (Set.memberOf currentMarks) state.priorMarks
+                                |> Set.toList
+
+                        startStyles =
+                            case startedMarks of
+                                [] ->
+                                    []
+
+                                _ ->
+                                    tagBeforeContent startedMarks :: List.concatMap .startStyles startedMarks
```

Comment:
🎨 super minor but we can get rid of the special treatment for `[]` here if we make `tagBeforeContent` return a List of styles rather than a single style (getting rid of the problematic `Css.batch []`).

Reply:
Good idea!